src/ceph-daemon/ceph-daemon

```diff
@@ -560,7 +565,7 @@ def extract_uid_gid():
 
 def deploy_daemon(fsid, daemon_type, daemon_id, c, uid, gid,
                   config, keyring):
-    # type: (str, str, Union[int, str], CephContainer, int, int, str, str) -> None
+    # type: (str, str, Union[int, str], CephContainer, int, int, Optional[str], Optional[str]) -> None
```

Comment:
Regarding

```
ceph-daemon:575: error: Argument 1 to "write" of "IO" has incompatible type "Optional[str]"; expected "str"
ceph-daemon:582: error: Argument 1 to "write" of "IO" has incompatible type "Optional[str]"; expected "str"
ceph-daemon:610: error: Argument 1 to "write" of "IO" has incompatible type "Optional[str]"; expected "str"
```

You cannot pass `None` to 

```python
575:        tmp_keyring.write(keyring)
```
See:

```
In [1]: f = open('/tmp/foo', 'w')

In [2]: f.write(None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-71806c76a6c6> in <module>()
----> 1 f.write(None)

TypeError: write() argument must be str, not None
```

So, either you make sure keyring cannot be `None` or you create the file only, if `keyring` is not None.

Reply:
If you really want to avoid dealing with it, you can simply remove the `# type:` line in this function.

Reply:
> Regarding
> 
> ```
> ceph-daemon:575: error: Argument 1 to "write" of "IO" has incompatible type "Optional[str]"; expected "str"
> ceph-daemon:582: error: Argument 1 to "write" of "IO" has incompatible type "Optional[str]"; expected "str"
> ceph-daemon:610: error: Argument 1 to "write" of "IO" has incompatible type "Optional[str]"; expected "str"
> ```
> 
> You cannot pass `None` to
> 
> ```python
> 575:        tmp_keyring.write(keyring)
> ```
> 
> See:
> 
> ```
> In [1]: f = open('/tmp/foo', 'w')
> 
> In [2]: f.write(None)
> ---------------------------------------------------------------------------
> TypeError                                 Traceback (most recent call last)
> <ipython-input-2-71806c76a6c6> in <module>()
> ----> 1 f.write(None)
> 
> TypeError: write() argument must be str, not None
> ```
> 
> So, either you make sure keyring cannot be `None` or you create the file only, if `keyring` is not None.

This seems like of those "should never happen" paths ... but I still think raising an exception when `None` would be helpful here (and it would fix the mypy errors). 

But, if we allow the create of an empty (or non-existent) keyring/config file, this could likely create a failure case that would be somewhat obscure to debug ...

Reply:
I just updated the condition to check that config and keyring aren't none and mypy is happy now

Reply:
assert works too